src/sleep/sleep.c

```diff
@@ -40,6 +43,161 @@
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (fiemap->fm_mapped_extents);
+
+                /* Resize fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(fiemap, fiemap_size,
+                                    sizeof(struct fiemap) + extents_size))
+                        return log_oom();
+
+                memzero(fiemap->fm_extents, extents_size);
+                fiemap->fm_extent_count = fiemap->fm_mapped_extents;
+                fiemap->fm_mapped_extents = 0;
+
+                if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0)
+                        return -errno;
+
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (result_extents +
+                                       fiemap->fm_mapped_extents);
+
+                /* Resize result_fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(result_fiemap, result_fiemap_size,
+                                    sizeof(struct fiemap) + extents_size))
+                        return log_oom();
+
+                memcpy(result_fiemap->fm_extents + result_extents,
+                       fiemap->fm_extents,
+                       sizeof(struct fiemap_extent) *
+                       fiemap->fm_mapped_extents);
+
+                result_extents += fiemap->fm_mapped_extents;
+
+                /* Highly unlikely that it is zero */
+                if (fiemap->fm_mapped_extents) {
+                        uint32_t i = fiemap->fm_mapped_extents - 1;
+
+                        fiemap_start = fiemap->fm_extents[i].fe_logical +
+                                       fiemap->fm_extents[i].fe_length;
+
+                        if (fiemap->fm_extents[i].fe_flags & FIEMAP_EXTENT_LAST)
+                                break;
+                }
+        }
+
+        result_fiemap->fm_mapped_extents = result_extents;
+        *ret = result_fiemap;
+        result_fiemap = NULL;
+        return 0;
+}
+
+static int write_disk_offset(void) {
+        _cleanup_free_ char *device = NULL, *type = NULL;
+        _cleanup_free_ struct fiemap *fiemap = NULL;
+        char offset_str[DECIMAL_STR_MAX(uint64_t)];
+        char device_str[DECIMAL_STR_MAX(uint64_t)];
+        struct stat stb;
+        uint64_t offset;
+        int fd;
+        int r;
+
+        r = find_hibernate_location(&device, &type, NULL, NULL);
+        if (r < 0)
+                return 0;
+
+        /* if it's a swap partition, we just write the disk to /sys/power/resume */
+        if (streq(type, "partition"))
+                return write_string_file("/sys/power/resume", device, 0);
+        else if (!streq(type, "file"))
+                return -EINVAL;
+
+        /* Only available in 4.17+ */
+        r = access("/sys/power/resume_offset", W_OK);
+        if (r < 0) {
+                log_debug_errno(r, "/sys/power/resume_offset unavailable: %m");
```

Comment:
the first parameter of log_debug_errno() should be the error code, but access() returns that in `errno`, not in its return value, hence this should be changed to: "log_debug_errno(errno, …"

Reply:
also, I'd prefer a more strict check, i.e. we should assume that resume_offset is unavailable only when access(F_OK) returns ENOENT. In all other cases we should consider this a proper error to propagate, since something else is wrong...